### ISSUE TITLE
osbuild-worker jobimpl-osbuild contains duplicate code and as a result osbuild-worker does not send compose status update to osbuild-composer

### ISSUE THREAD

**Issue by msehnout** [osbuild/osbuild-composer#1276]
https://github.com/osbuild/osbuild-composer/blob/167fdc44bc82744b759a059b197d4415523e83da/cmd/osbuild-worker/jobimpl-osbuild.go#L101

Almost the same piece of code lives in the Run function twice. This is the first one:
```go
	if osbuildOutput.Success && args.ImageName != "" {
		var f *os.File
		imagePath := path.Join(outputDirectory, args.ImageName)
		if args.StreamOptimized {
			f, err = vmware.OpenAsStreamOptimizedVmdk(imagePath)
			if err != nil {
				return err
			}
			streamOptimizedPath = f.Name()
		} else {
			f, err = os.Open(imagePath)
			if err != nil {
				return err
			}
		}
		err = job.UploadArtifact(args.ImageName, f)
		if err != nil {
			return err
		}
	}
```
and this is the second one:
```go
			if !osbuildOutput.Success {
				continue
			}
			var f *os.File
			imagePath := path.Join(outputDirectory, options.Filename)
			if options.StreamOptimized {
				f, err = vmware.OpenAsStreamOptimizedVmdk(imagePath)
				if err != nil {
					r = append(r, err)
					continue
				}
			} else {
				f, err = os.Open(imagePath)
				if err != nil {
					r = append(r, err)
					continue
				}
			}
```

Now the problem is that the first one contains the `return` statement which makes the function end without calling `job.Update` at the end:
https://github.com/osbuild/osbuild-composer/blob/167fdc44bc82744b759a059b197d4415523e83da/cmd/osbuild-worker/jobimpl-osbuild.go#L545

Now this causes osbuild-worker to correctly log the fact that a compose failed:
```
Mar 08 09:39:00 vm osbuild-worker[26301]: Running 'osbuild' job 24b774cf-4ae0-49b2-8f35-cdcd17c83763
Mar 08 09:43:52 vm osbuild-worker[26301]: 2021/03/08 09:43:52 Job 24b774cf-4ae0-49b2-8f35-cdcd17c83763 failed: open /var/cache/osbuild-worker/output/24b774cf-4ae0-49b2-8f35-cdcd17c83763-803191209/commit.tar: no such file or directory
Mar 08 09:43:52 vm osbuild-worker[26301]: Waiting for a new job...
```
But osbuild-composer does not know about it:
```
[root@vm ~]# composer-cli compose status
b620987c-25ba-4e02-9ddc-1dbcf5f543f0 RUNNING  Mon Mar  8 09:00:57 2021 uid-bug-upgrade 0.1.0 rhel-edge-commit 
4f723253-c38f-4758-9751-833b0b1e948c RUNNING  Mon Mar  8 09:39:00 2021 uid-bug-upgrade 0.1.0 rhel-edge-commit 
9aa538b1-c1ac-464e-b739-a96c787d4091 FINISHED Mon Mar  8 08:24:15 2021 uid-bug         0.1.0 rhel-edge-commit 
```

It looks like the first occurence is there by accident.

**Comment by teg** [comment]
These two chunks of code should be unified, but they actually do slightly different things, so can't just drop one:

 - the first one converts the VMDK image before uploading back to composer (only used when using the weldr API)
 - the second converts the VMDK when uploading directly to vSphere (which could be used from weldr, or (in the future) independently).

It would be nice to the the conversion only once, but note that currently any combination of the two code blocks are possible (either, none, or both). Probably we should just look at `args.StreamOptimized` and convert the image if that is set. Now I see we support setting it also in the target options, but maybe we should treat this more as a property of the manifest (which is what it should be once osbuild gains support for that), and ignore the per-target option. What do you think @jmikovic?

As to the missing error handling: thoughts on this @larskarlitski? We can fail to do the image conversion (internal error, I'd say?) or to call the `UploadArtifact()` call, but in the latter case are we any more likely to succeed with the `Update()` call on the same API?

**Comment by msehnout** [comment]
> As to the missing error handling: thoughts on this @larskarlitski? We can fail to do the image conversion (internal error, I'd say?) or to call the `UploadArtifact()` call, but in the latter case are we any more likely to succeed with the `Update()` call on the same API?

We can also fail while trying to open the file (`f, err = os.Open(imagePath)`) as you can see in the output above. Another question is why the file is not there, but I haven't been able to answer that yet :-)

**Comment by teg** [comment]
D'oh, I didn't do the horizontal scroll... Could you check the journal on the machine you see that issue? I've only seen this with i/o errors in the kernel log, blaming it on PSI storage being bad. Seems suspicious that it keeps on failing on the same spot though...

**Comment by msehnout** [comment]
First run:
```
Mar 08 09:00:57 vm osbuild-worker[26301]: Running 'osbuild' job 369af605-2197-4f11-b1b5-d2c2fbfc51f2
Mar 08 09:01:01 vm CROND[56628]: (root) CMD (run-parts /etc/cron.hourly)
Mar 08 09:01:01 vm run-parts[56631]: (/etc/cron.hourly) starting 0anacron
Mar 08 09:01:01 vm anacron[56644]: Anacron started on 2021-03-08
Mar 08 09:01:01 vm run-parts[56646]: (/etc/cron.hourly) finished 0anacron
Mar 08 09:01:01 vm anacron[56644]: Will run job `cron.daily' in 21 min.
Mar 08 09:01:01 vm anacron[56644]: Will run job `cron.weekly' in 41 min.
Mar 08 09:01:01 vm anacron[56644]: Will run job `cron.monthly' in 61 min.
Mar 08 09:01:01 vm anacron[56644]: Jobs will be executed sequentially
Mar 08 09:02:06 vm systemd-udevd[57829]: Using default interface naming scheme 'rhel-8.0'.
Mar 08 09:02:06 vm systemd-udevd[57828]: Using default interface naming scheme 'rhel-8.0'.
Mar 08 09:04:15 vm dbus-daemon[704]: [system] Activating service name='org.fedoraproject.Setroubleshootd' requested by ':1.44' (uid=0 pid=676 comm="/usr/sbin/sedispatch " label="system_u:system_r:auditd_t:s0") (using servicehelper)
Mar 08 09:04:15 vm dbus-daemon[68994]: [system] Failed to reset fd limit before activating service: org.freedesktop.DBus.Error.AccessDenied: Failed to restore old fd limit: Operation not permitted
Mar 08 09:04:16 vm dbus-daemon[704]: [system] Successfully activated service 'org.fedoraproject.Setroubleshootd'
Mar 08 09:04:18 vm setroubleshoot[68994]: AnalyzeThread.run(): Cancel pending alarm
Mar 08 09:04:18 vm dbus-daemon[704]: [system] Activating service name='org.fedoraproject.SetroubleshootPrivileged' requested by ':1.138' (uid=995 pid=68994 comm="/usr/libexec/platform-python -Es /usr/sbin/setroub" label="system_u:system_r:setroubleshootd_t:s0-s0:c0.c1023") (using servicehelper)
Mar 08 09:04:18 vm dbus-daemon[70264]: [system] Failed to reset fd limit before activating service: org.freedesktop.DBus.Error.AccessDenied: Failed to restore old fd limit: Operation not permitted
Mar 08 09:04:18 vm dbus-daemon[704]: [system] Successfully activated service 'org.fedoraproject.SetroubleshootPrivileged'
Mar 08 09:04:19 vm setroubleshoot[68994]: SELinux is preventing semodule from using the 'nnp_transition, nosuid_transition' accesses on a process. For complete SELinux messages run: sealert -l 538f04ae-6611-438a-93d3-f1f284a7a8be
Mar 08 09:04:19 vm setroubleshoot[68994]: SELinux is preventing semodule from using the 'nnp_transition, nosuid_transition' accesses on a process.
                                          
                                          *****  Plugin catchall (100. confidence) suggests   **************************
                                          
                                          If you believe that semodule should be allowed nnp_transition nosuid_transition access on processes labeled setfiles_mac_t by default.
                                          Then you should report this as a bug.
                                          You can generate a local policy module to allow this access.
                                          Do
                                          allow this access for now by executing:
                                          # ausearch -c 'semodule' --raw | audit2allow -M my-semodule
                                          # semodule -X 300 -i my-semodule.pp
                                          
Mar 08 09:04:19 vm setroubleshoot[68994]: AnalyzeThread.run(): Set alarm timeout to 10
Mar 08 09:04:57 vm dbus-daemon[704]: [system] Activating service name='org.fedoraproject.Setroubleshootd' requested by ':1.44' (uid=0 pid=676 comm="/usr/sbin/sedispatch " label="system_u:system_r:auditd_t:s0") (using servicehelper)
Mar 08 09:04:57 vm dbus-daemon[79106]: [system] Failed to reset fd limit before activating service: org.freedesktop.DBus.Error.AccessDenied: Failed to restore old fd limit: Operation not permitted
Mar 08 09:05:00 vm dbus-daemon[704]: [system] Successfully activated service 'org.fedoraproject.Setroubleshootd'
Mar 08 09:05:01 vm setroubleshoot[79106]: AnalyzeThread.run(): Cancel pending alarm
Mar 08 09:05:01 vm dbus-daemon[704]: [system] Activating service name='org.fedoraproject.SetroubleshootPrivileged' requested by ':1.143' (uid=995 pid=79106 comm="/usr/libexec/platform-python -Es /usr/sbin/setroub" label="system_u:system_r:setroubleshootd_t:s0-s0:c0.c1023") (using servicehelper)
Mar 08 09:05:01 vm dbus-daemon[79117]: [system] Failed to reset fd limit before activating service: org.freedesktop.DBus.Error.AccessDenied: Failed to restore old fd limit: Operation not permitted
Mar 08 09:05:02 vm dbus-daemon[704]: [system] Successfully activated service 'org.fedoraproject.SetroubleshootPrivileged'
Mar 08 09:05:04 vm setroubleshoot[79106]: SELinux is preventing semodule from using the 'nnp_transition, nosuid_transition' accesses on a process. For complete SELinux messages run: sealert -l 538f04ae-6611-438a-93d3-f1f284a7a8be
Mar 08 09:05:04 vm setroubleshoot[79106]: SELinux is preventing semodule from using the 'nnp_transition, nosuid_transition' accesses on a process.
                                          
                                          *****  Plugin catchall (100. confidence) suggests   **************************
                                          
                                          If you believe that semodule should be allowed nnp_transition nosuid_transition access on processes labeled setfiles_mac_t by default.
                                          Then you should report this as a bug.
                                          You can generate a local policy module to allow this access.
                                          Do
                                          allow this access for now by executing:
                                          # ausearch -c 'semodule' --raw | audit2allow -M my-semodule
                                          # semodule -X 300 -i my-semodule.pp
                                          
Mar 08 09:05:04 vm setroubleshoot[79106]: AnalyzeThread.run(): Set alarm timeout to 10
Mar 08 09:06:15 vm osbuild-worker[26301]: 2021/03/08 09:06:15 Job 369af605-2197-4f11-b1b5-d2c2fbfc51f2 failed: open /var/cache/osbuild-worker/output/369af605-2197-4f11-b1b5-d2c2fbfc51f2-112330482/commit.tar: no such file or directory
Mar 08 09:06:15 vm osbuild-worker[26301]: Waiting for a new job...
```
In between runs:
```
Mar 08 09:09:13 vm chronyd[736]: Selected source 49.12.125.53
Mar 08 09:22:01 vm anacron[56644]: Job `cron.daily' started
Mar 08 09:22:01 vm run-parts[79167]: (/etc/cron.daily) starting logrotate
Mar 08 09:22:01 vm run-parts[79172]: (/etc/cron.daily) finished logrotate
Mar 08 09:22:01 vm anacron[56644]: Job `cron.daily' terminated
Mar 08 09:38:59 vm sudo[79189]:    admin : TTY=pts/1 ; PWD=/root ; USER=root ; COMMAND=/bin/curl --header Content-Type: application/json --unix-socket /run/weldr/api.socket http://localhost/api/v1/compose --data @upgrade-request.json
Mar 08 09:38:59 vm sudo[79189]: pam_systemd(sudo:session): Cannot create session: Already running in a session or user slice
Mar 08 09:38:59 vm sudo[79189]: pam_unix(sudo:session): session opened for user root by admin(uid=0)
Mar 08 09:39:00 vm sudo[79189]: pam_unix(sudo:session): session closed for user root
```
Second run:
```
Mar 08 09:39:00 vm osbuild-worker[26301]: Running 'osbuild' job 24b774cf-4ae0-49b2-8f35-cdcd17c83763
Mar 08 09:39:59 vm systemd-udevd[80637]: Using default interface naming scheme 'rhel-8.0'.
Mar 08 09:39:59 vm systemd-udevd[80638]: Using default interface naming scheme 'rhel-8.0'.
Mar 08 09:42:01 vm anacron[56644]: Job `cron.weekly' started
Mar 08 09:42:01 vm anacron[56644]: Job `cron.weekly' terminated
Mar 08 09:42:07 vm dbus-daemon[704]: [system] Activating service name='org.fedoraproject.Setroubleshootd' requested by ':1.44' (uid=0 pid=676 comm="/usr/sbin/sedispatch " label="system_u:system_r:auditd_t:s0") (using servicehelper)
Mar 08 09:42:07 vm dbus-daemon[91805]: [system] Failed to reset fd limit before activating service: org.freedesktop.DBus.Error.AccessDenied: Failed to restore old fd limit: Operation not permitted
Mar 08 09:42:08 vm dbus-daemon[704]: [system] Successfully activated service 'org.fedoraproject.Setroubleshootd'
Mar 08 09:42:09 vm setroubleshoot[91805]: AnalyzeThread.run(): Cancel pending alarm
Mar 08 09:42:09 vm dbus-daemon[704]: [system] Activating service name='org.fedoraproject.SetroubleshootPrivileged' requested by ':1.150' (uid=995 pid=91805 comm="/usr/libexec/platform-python -Es /usr/sbin/setroub" label="system_u:system_r:setroubleshootd_t:s0-s0:c0.c1023") (using servicehelper)
Mar 08 09:42:09 vm dbus-daemon[92386]: [system] Failed to reset fd limit before activating service: org.freedesktop.DBus.Error.AccessDenied: Failed to restore old fd limit: Operation not permitted
Mar 08 09:42:09 vm dbus-daemon[704]: [system] Successfully activated service 'org.fedoraproject.SetroubleshootPrivileged'
Mar 08 09:42:10 vm setroubleshoot[91805]: SELinux is preventing semodule from using the 'nnp_transition, nosuid_transition' accesses on a process. For complete SELinux messages run: sealert -l 538f04ae-6611-438a-93d3-f1f284a7a8be
Mar 08 09:42:10 vm setroubleshoot[91805]: SELinux is preventing semodule from using the 'nnp_transition, nosuid_transition' accesses on a process.
                                          
                                          *****  Plugin catchall (100. confidence) suggests   **************************
                                          
                                          If you believe that semodule should be allowed nnp_transition nosuid_transition access on processes labeled setfiles_mac_t by default.
                                          Then you should report this as a bug.
                                          You can generate a local policy module to allow this access.
                                          Do
                                          allow this access for now by executing:
                                          # ausearch -c 'semodule' --raw | audit2allow -M my-semodule
                                          # semodule -X 300 -i my-semodule.pp
                                          
Mar 08 09:42:10 vm setroubleshoot[91805]: AnalyzeThread.run(): Set alarm timeout to 10
Mar 08 09:42:47 vm dbus-daemon[704]: [system] Activating service name='org.fedoraproject.Setroubleshootd' requested by ':1.44' (uid=0 pid=676 comm="/usr/sbin/sedispatch " label="system_u:system_r:auditd_t:s0") (using servicehelper)
Mar 08 09:42:47 vm dbus-daemon[101915]: [system] Failed to reset fd limit before activating service: org.freedesktop.DBus.Error.AccessDenied: Failed to restore old fd limit: Operation not permitted
Mar 08 09:42:49 vm dbus-daemon[704]: [system] Successfully activated service 'org.fedoraproject.Setroubleshootd'
Mar 08 09:42:50 vm setroubleshoot[101915]: AnalyzeThread.run(): Cancel pending alarm
Mar 08 09:42:50 vm dbus-daemon[704]: [system] Activating service name='org.fedoraproject.SetroubleshootPrivileged' requested by ':1.155' (uid=995 pid=101915 comm="/usr/libexec/platform-python -Es /usr/sbin/setroub" label="system_u:system_r:setroubleshootd_t:s0-s0:c0.c1023") (using servicehelper)
Mar 08 09:42:50 vm dbus-daemon[101926]: [system] Failed to reset fd limit before activating service: org.freedesktop.DBus.Error.AccessDenied: Failed to restore old fd limit: Operation not permitted
Mar 08 09:42:50 vm dbus-daemon[704]: [system] Successfully activated service 'org.fedoraproject.SetroubleshootPrivileged'
Mar 08 09:42:51 vm setroubleshoot[101915]: SELinux is preventing semodule from using the 'nnp_transition, nosuid_transition' accesses on a process. For complete SELinux messages run: sealert -l 538f04ae-6611-438a-93d3-f1f284a7a8be
Mar 08 09:42:51 vm setroubleshoot[101915]: SELinux is preventing semodule from using the 'nnp_transition, nosuid_transition' accesses on a process.
                                           
                                           *****  Plugin catchall (100. confidence) suggests   **************************
                                           
                                           If you believe that semodule should be allowed nnp_transition nosuid_transition access on processes labeled setfiles_mac_t by default.
                                           Then you should report this as a bug.
                                           You can generate a local policy module to allow this access.
                                           Do
                                           allow this access for now by executing:
                                           # ausearch -c 'semodule' --raw | audit2allow -M my-semodule
                                           # semodule -X 300 -i my-semodule.pp
                                           
Mar 08 09:42:51 vm setroubleshoot[101915]: AnalyzeThread.run(): Set alarm timeout to 10
Mar 08 09:43:52 vm osbuild-worker[26301]: 2021/03/08 09:43:52 Job 24b774cf-4ae0-49b2-8f35-cdcd17c83763 failed: open /var/cache/osbuild-worker/output/24b774cf-4ae0-49b2-8f35-cdcd17c83763-803191209/commit.tar: no such file or directory
Mar 08 09:43:52 vm osbuild-worker[26301]: Waiting for a new job...
```

**Comment by jmikovic** [comment]
> It would be nice to the the conversion only once, but note that currently any combination of the two code blocks are possible (either, none, or both). Probably we should just look at `args.StreamOptimized` and convert the image if that is set. Now I see we support setting it also in the target options, but maybe we should treat this more as a property of the manifest (which is what it should be once osbuild gains support for that), and ignore the per-target option. What do you think @jmikovic?

I'm definitely in favour of only once and unifying `StreamOptimized` options to one place.
I was actually looking into this when adding upload target for VMWare but local upload target takes the path from `options.Filename` instead of `args.ImageName` so I decided to leave it. AFAIK these variables should contain the same values but it's something that should be cleaned up as well.

**Comment by msehnout** [comment]
Update on this issue:

I've actually stumbled upon 2 different bug. First one is in osbuild and the problem is that osbuild reports success even in case the assembler failed (in this case because of insufficient hdd capacity):
 * osbuild output from the failed run
https://gist.githubusercontent.com/msehnout/403dfb3045377a44f730812867ba9204/raw/e69e6f2196ffc42a5af7f5ebc8ba6c2a57ff8b16/gistfile0.txt

This has been promptly fixed by @gicmo :1st_place_medal: :
https://github.com/osbuild/osbuild/pull/606

The second bug is in osbuild-worker and the problem is that errors are not reported to the worker API. My attempt on fixing it is the PR #1281 .

As far as the `StreamOptimized` option is concerned, I'd create a new issue for it and track it there. This one is already convoluted enough :-)
